File: hospital-backend/app/common/datetime/parser.py
"""Parse timestamp strings to datetime objects."""

from datetime import datetime, timezone
from typing import Union


def parse_iso8601(timestamp_str: str) -> datetime:
    """
    Parse ISO8601 string to UTC datetime (timezone-aware).

    Handles both 'Z' suffix and explicit timezone offsets.

    Args:
        timestamp_str: ISO8601 formatted string (e.g., "2025-11-10T14:30:00Z")

    Returns:
        Timezone-aware datetime in UTC

    Example:
        >>> parse_iso8601("2025-11-10T14:30:00Z")
        datetime(2025, 11, 10, 14, 30, 0, tzinfo=timezone.utc)
    """
    # Handle 'Z' suffix (Zulu time = UTC)
    if timestamp_str.endswith('Z'):
        timestamp_str = timestamp_str.replace('Z', '+00:00')

    dt = datetime.fromisoformat(timestamp_str)

    # Ensure timezone-aware
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)

    return dt


def parse_timestamp(value: Union[str, datetime, int, float]) -> datetime:
    """
    Parse various timestamp formats to datetime.

    Args:
        value: ISO8601 string, datetime object, or Unix epoch (int/float)

    Returns:
        Timezone-aware datetime in UTC

    Example:
        >>> parse_timestamp("2025-11-10T14:30:00Z")
        >>> parse_timestamp(1699627800)
        >>> parse_timestamp(datetime.now())
    """
    if isinstance(value, datetime):
        if value.tzinfo is None:
            return value.replace(tzinfo=timezone.utc)
        return value

    if isinstance(value, str):
        return parse_iso8601(value)

    if isinstance(value, (int, float)):
        return datetime.fromtimestamp(value, tz=timezone.utc)

    raise ValueError(f"Cannot parse timestamp from type {type(value)}")

File: hospital-display-app/src/utils/datetime/parser.ts
/**
 * Parse timestamp strings to Date objects.
 */

export function parseISO(isoString: string): Date {
  /**
   * Parse ISO8601 string to Date object.
   *
   * @param isoString - ISO8601 formatted string
   * @returns Date object
   *
   * @example
   * parseISO("2025-11-10T14:30:00Z")
   */
  return new Date(isoString);
}

export function parseTimestamp(value: string | number | Date): Date {
  /**
   * Parse various timestamp formats to Date.
   *
   * @param value - ISO string, Unix epoch, or Date object
   * @returns Date object
   *
   * @example
   * parseTimestamp("2025-11-10T14:30:00Z")
   * parseTimestamp(1699627800)
   * parseTimestamp(new Date())
   */
  if (value instanceof Date) {
    return value;
  }

  if (typeof value === 'string') {
    return parseISO(value);
  }

  if (typeof value === 'number') {
    return new Date(value * 1000); // Assume Unix epoch in seconds
  }

  throw new Error(`Cannot parse timestamp from type ${typeof value}`);
}
